### PR TITLE
RHV Provider Role and Playbooks

### PR DESCRIPTION
**Pull request by cwilkers**
Adds playbook and role for creating OpenShift VMs on a RHV cluster

### CODE REVIEW

File: roles/openshift_rhv/defaults/main.yml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+      storage_domain: "{{ openshift_rhv_data_store }}"
+      name: docker_disk
+      interface: virtio
+    - size: 30GiB
+      storage_domain: "{{ openshift_rhv_data_store }}"
+      name: localvol_disk
+      interface: virtio
+    state: running
+
+openshift_rhv_vm_manifest:
```
Comment (sradco) [inline roles/openshift_rhv/defaults/main.yml]: I think this is too big for a default installation. I think we should have the default based on https://docs.openshift.com/container-platform/3.10/install/index.html#single-master-multi-node

and either have the additional nodes commented out and reference the OpenShift documentation https://docs.openshift.com/container-platform/3.10/install/index.html#environment-scenarios or add additional defaults.
- Single Master and Node on One System
- Single Master and Multiple Nodes
- Single Master, Multiple etcd, and Multiple Nodes
- Multiple Masters Using Native HA with Co-located Clustered etcd
- Multiple Masters Using Native HA with External Clustered etcd

We need to support  an all-in-one installation if only 1 master node is chosen, for the metrics store.
That will require adding the vm profile that is similler to the master profile but with bigger disk for elasticsearch storage.


 @cwilkers - Can you please share the playbook that will deploy OpenShift?

File: playbooks/rhv/README.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
+# RHV Playbooks
+## Provisioning
+This subdirectory contains the Ansible playbooks used to deploy 
+an OpenShift Container Platform environment on Red Hat Virtualization
+
+### Where do I start?
+Choose a host from which Ansible plays will be executed. This host must have
+the ability to access the web interface of the RHV cluster engine and the
+network on which the OpenShift nodes will be installed. We will refer to
+this host as the *bastion*.
```
Comment (sradco) [inline playbooks/rhv/README.md]: Why not use the engine machine for the deployment?
This will simplify the vars setting for the users.

File: playbooks/rhv/provisioning-vars.yaml.example [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+openshift_rhv_data_store: # vmstore
+openshift_rhv_ssh_key: "{{ lookup('file', '~/.ssh/id_rsa.pub') }}"
+
+##########################
+# Template Creation
+# https://github.com/oVirt/ovirt-ansible-image-template
+##########################
+ovirt_qcow_url: # https://cloud.centos.org/centos/7/images/CentOS-7-x86_64-GenericCloud.qcow2c
+ovirt_image_path: "{{ lookup('env', 'HOME') }}/Downloads/{{ ovirt_template_name }}.qcow2"
+ovirt_template_name: # rhel75
```
Comment (sradco) [inline playbooks/rhv/provisioning-vars.yaml.example]: Why not set a default here?

File: roles/openshift_rhv/defaults/main.yml [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+- name: 'master'
```
[the remark this answers is not in the document]
Comment (jirib) [reply roles/openshift_rhv/defaults/main.yml]: This is example and as such you can change it with `sed` in one simple shot.

File: playbooks/rhv/README.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
+# RHV Playbooks
```
Comment (sradco) [inline playbooks/rhv/README.md]: I think these should be "oVirt" playbooks the same as ovirt-ansible playbooks.
And all the RHV references should be converted to ovirt.
In the OCP documentation the name can be RHV but these roles are also meant to work upstream. No?

File: roles/openshift_rhv/defaults/main.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (sradco) [reply roles/openshift_rhv/defaults/main.yml]: I disagree...
Defaults are meant so that the playbook will be able to run with no special intervention.
I think using sed is a workaround and not really how users would like to work.
You expect the users to know your playbook and start messing with it... 

If you give them a parameter like "vm_name_prefix" with default like openshift.
And add a comment that this in a name they need to choose to identify their OpenShift installation(It can be a staging/prod/test/ customers name etc.)and  will allow the creation of different vms for several OpenShift installations
will make it more structured and less work for the users.

File: playbooks/rhv/README.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+with values from your RHV environment, making sure to fill in all commented values.
+
+*Red Hat Virtualization Certificate*
+
+A copy of the `/etc/pki/ovirt-engine/ca.pem` from the RHV engine will need to
+be added to the `reference-architecture/rhv-ansible` directory. Replace the
+example server in the following command to download the certificate:
+
+```
+$ curl --output ca.pem 'http://engine.example.com/ovirt-engine/services/pki-resource?resource=ca-certificate&format=X509-PEM-CA'
```
Comment (sradco) [inline playbooks/rhv/README.md]: Why not add this to the playbook and fetch it with the engine_url, user and password?

File: roles/openshift_rhv/tasks/build_vm_list.yml [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+      {{ vms|default([]) }} + [
+      {% for iter in range(item.count) -%}
+      {% if iter > 0 -%},{% endif -%}
+      {
+      'name': '{{ item.name }}{{ iter }}.{{ public_hosted_zone }}',
+      'tag': 'openshift_{{ item.profile }}',
+      'cloud_init':
+      {
+      'host_name': '{{ item.name }}{{ iter }}.{{ public_hosted_zone }}',
+      'authorized_ssh_keys': '{{ openshift_rhv_ssh_key }}'
```
Comment (sradco) [inline roles/openshift_rhv/tasks/build_vm_list.yml]: Can be added to the [all:vars]

Like:
[all:vars]
ansible_ssh_private_key_file = '~/.ssh/id_rsa.pub'

File: roles/openshift_rhv/defaults/main.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (cwilkers) [reply roles/openshift_rhv/defaults/main.yml]: This is do-able, could you provide a patch? Any variables we add should be prefixed with `openshift_rhv_` I would have the default as the empty string.

File: roles/openshift_rhv/defaults/main.yml (hunk 1 above)
Comment (cwilkers) [reply roles/openshift_rhv/defaults/main.yml]: ../prerequisites.yml and ../deploy-cluster.yml

File: playbooks/rhv/README.md (hunk 2 above)
Comment (cwilkers) [reply playbooks/rhv/README.md]: I disagree with directing anybody to run commands on the engine vm. The goal of simplifying for the users is better served by showing all the variables needed for the general case.

File: roles/openshift_rhv/defaults/main.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (sradco) [reply roles/openshift_rhv/defaults/main.yml]: Yes. I'll can do that.

File: roles/openshift_rhv/defaults/main.yml (hunk 1 above)
Comment (sradco) [reply roles/openshift_rhv/defaults/main.yml]: I thought you are working on generating the openshift inventory file based on the nodes you created.
I generate the openshift inventory file in my playbook, based on the given nodes so I can tailor this together. I only need to check with you about the parameters that I'm missing from my files.

File: playbooks/rhv/README.md (hunk 2 above)
Comment (sradco) [reply playbooks/rhv/README.md]: +1

File: playbooks/rhv/inventory.example [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+masters
+etcd
+lb
+
+[OSEv3:vars]
+# General variables
+ansible_ssh_user=root
+console_port=8443
+debug_level=2
+deployment_type=openshift-enterprise
```
Comment (sradco) [inline playbooks/rhv/inventory.example]: Should be origin or okd in upstream.
I added a vars file that overrides this in the d/s rpm.

File: playbooks/rhv/inventory.example [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+openshift_debug_level="{{ debug_level }}"
+openshift_deployment_type="{{ deployment_type }}"
+openshift_master_cluster_method=native
+openshift_node_debug_level="{{ node_debug_level | default(debug_level, true) }}"
+openshift_release='3.10'
+openshift_vers=v3_10
+
+# RHV Specific Settings
```
Comment (sradco) [inline playbooks/rhv/inventory.example]: I think this section should be added to https://github.com/openshift/openshift-ansible/blob/master/inventory/hosts.example

File: playbooks/rhv/inventory.example [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+# openshift_cloudprovider_ovirt_password=
+
+# Pod Networking
+os_sdn_network_plugin_name=redhat/openshift-ovs-networkpolicy
+
+# Registry
+openshift_hosted_registry_replicas=1
+openshift_hosted_registry_storage_kind=nfs
+openshift_hosted_registry_storage_access_modes=['ReadWriteMany']
+openshift_hosted_registry_storage_host=
```
Comment (sradco) [inline playbooks/rhv/inventory.example]: Should be openshift_hosted_registry_storage_host=nfs..{{public_hosted_zone}} no?

Comment (sradco) [reply playbooks/rhv/inventory.example]: Except for this parameter why are the other required here ? Why not refer to the https://github.com/openshift/openshift-ansible/blob/master/inventory/hosts.example for all the defaults and options? Its confusing...

File: roles/openshift_rhv/defaults/main.yml (hunk 1 above)
Comment (cwilkers) [reply roles/openshift_rhv/defaults/main.yml]: I'm not generating an inventory file, just creating the vms data structure and, when I have DNS again, I can add IPs to the inventory hosts' variables.

File: playbooks/rhv/README.md (hunk 5 above)
Comment (cwilkers) [reply playbooks/rhv/README.md]: I'm okay with doing this. It should likely be a wider conversation first, though.
CC @cooktheryan @sdodson

File: playbooks/rhv/README.md (hunk 6 above)
Comment (cwilkers) [reply playbooks/rhv/README.md]: I toyed with the idea, but then that's equivalent to setting the insecure flag on auth. At least with a manual step, the user has a chance to verify the right CA is being pulled.

File: playbooks/rhv/inventory.example (hunk 9 above)
Comment (cwilkers) [reply playbooks/rhv/inventory.example]: It should probably be removed until we have a cloud provider ready to go on install.

File: playbooks/rhv/inventory.example (hunk 10 above)
Comment (cwilkers) [reply playbooks/rhv/inventory.example]: This is included in error... I had removed the hosted registry lines at one point, but may have lost those changes. I'll take another look at streamlining the example inventory.

File: playbooks/rhv/inventory.example (hunk 9 above)
Comment (sradco) [reply playbooks/rhv/inventory.example]: I think we should remove as much as we can and leave the oVirt / RHV specifics.
And reference the general variables to the "hosts.example" file for more options.
And if there are variables that are rhv specific variable with defaults that were added to the core openshift-playbooks(deploy_cluster, prerequisites), they should be added to the  "hosts.example" file.

File: playbooks/rhv/README.md (hunk 6 above)
Comment (sradco) [reply playbooks/rhv/README.md]: I generated the engine ca and then copied it. Will that be more secure? https://gerrit.ovirt.org/#/c/93341/23/roles/oVirt.add-metrics-store-host/tasks/main.yml

File: playbooks/rhv/inventory.example (hunk 10 above)
Comment (sradco) [reply playbooks/rhv/inventory.example]: +1

File: playbooks/rhv/inventory.example [hunk 11]
```diff
@@ -0,0 +1,99 @@
+[all:vars]
```
Comment (sradco) [inline playbooks/rhv/inventory.example]: Based on https://docs.openshift.com/container-platform/3.10/install/prerequisites.html#install-config-network-using-firewalld, I think we may want to add the flag os_firewall_use_firewalld=true
@cwilkers

File: playbooks/rhv/README.md (hunk 6 above)
Comment (sradco) [reply playbooks/rhv/README.md]: @cwilkers

File: playbooks/rhv/inventory.example [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+[OSEv3:children]
+nodes
+masters
+etcd
+lb
+
+[OSEv3:vars]
+# General variables
+ansible_ssh_user=root
```
Comment (sradco) [inline playbooks/rhv/inventory.example]: Please add to the vars in parentheses when that is the default value.

Comment (sradco) [reply playbooks/rhv/inventory.example]: Like: ansible_ssh_user=root (Default)

Comment (sradco) [reply playbooks/rhv/inventory.example]: Or just have them commented out
like:
# ansible_ssh_user=root

File: playbooks/rhv/inventory.example [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+console_port=8443
+debug_level=2
+deployment_type=openshift-enterprise
+openshift_debug_level="{{ debug_level }}"
+openshift_deployment_type="{{ deployment_type }}"
+openshift_master_cluster_method=native
+openshift_node_debug_level="{{ node_debug_level | default(debug_level, true) }}"
+openshift_release='3.10'
```
Comment (sradco) [inline playbooks/rhv/inventory.example]: In my inventory file I set it like this:
openshift_release=v3.10

It works for me.
Does '3.10' work for you?

I also have:
short_version=3.10

File: playbooks/rhv/inventory.example (hunk 11 above)
Comment (sradco) [inline playbooks/rhv/inventory.example]: I needed to add here
ansible_ssh_private_key_file = /path/to/engine_id_rsa

for the passwordless connection.

File: playbooks/rhv/README.md [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+#### Step 2 RHV Provisioning Variables
+
+Fill out a provisioning variables file (example [`provisioning-vars.yaml.example`](provisioning-vars.yaml.example)
+with values from your RHV environment, making sure to fill in all commented values.
+
+*Red Hat Virtualization Certificate*
+
+A copy of the `/etc/pki/ovirt-engine/ca.pem` from the RHV engine will need to
+be added to the `reference-architecture/rhv-ansible` directory. Replace the
+example server in the following command to download the certificate:
```
Comment (sradco) [inline playbooks/rhv/README.md]: Isn't the ca.pem fetched by the ovirt playbooks?
We have the engine url/ user and password...
@machacekondra @cwilkers @rgolangh

File: playbooks/rhv/inventory.example [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+openshift_hosted_registry_storage_kind=nfs
+openshift_hosted_registry_storage_access_modes=['ReadWriteMany']
+openshift_hosted_registry_storage_host=
+openshift_hosted_registry_storage_nfs_directory=/var/lib/exports
+openshift_hosted_registry_storage_volume_name=registryvol
+openshift_hosted_registry_storage_volume_size=20Gi
+openshift_enable_unsupported_configurations=True
+
+# Authentication
+openshift_master_identity_providers="[{'name': 'htpasswd_auth', 'login': 'True', 'challenge': 'True', 'kind': 'HTPasswdPasswordIdentityProvider', 'filename': '/etc/origin/master/htpasswd'}]"
```
Comment (sradco) [inline playbooks/rhv/inventory.example]: In my inventory file it is set to:
openshift_master_identity_providers=[{'mappingMethod': 'lookup', 'challenge': 'true', 'login': 'true', 'kind': 'AllowAllPasswordIdentityProvider', 'name': 'allow_all'}]

Can you please help me figure out the difference?

File: playbooks/rhv/provisioning-vars.yaml.example [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+# Virtual machine profile for master nodes
+# Differs from node profile by 16GiB RAM, extra disk for etcd
+master_vm:
+  cluster: "{{ openshift_rhv_cluster }}"
+  template: "{{ template_name }}"
+  memory: 16GiB
+  cores: 2
+  high_availability: true
+  disks:
+    - size: 15GiB
```
Comment (sradco) [inline playbooks/rhv/provisioning-vars.yaml.example]: How are the disks mounted to the vm?
To what path?

File: playbooks/rhv/openshift-cluster/register-nodes.yml [hunk 17]
```diff
@@ -0,0 +1,12 @@
+---
+- name: Subscribe nodes
+  hosts:
+    - nodes
+    - lb
```
Comment (sradco) [inline playbooks/rhv/openshift-cluster/register-nodes.yml]: Please add "if lb defined" in case user does not request lb

File: playbooks/rhv/inventory.example (hunk 12 above)
Comment (sradco) [reply playbooks/rhv/inventory.example]: Also, I see that it was deprecated in favor of ansible_user

File: playbooks/rhv/inventory.example [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+[OSEv3:children]
+nodes
+masters
+etcd
+lb
+
+[OSEv3:vars]
+# General variables
+ansible_ssh_user=root
+console_port=8443
```
Comment (sradco) [inline playbooks/rhv/inventory.example]: Amazing... no documentation that I could find about console_port.
But did find it in the code
https://github.com/openshift/openshift-ansible/blob/master/roles/openshift_facts/library/openshift_facts.py

8443 seems to be the default value

Comment (sradco) [inline playbooks/rhv/inventory.example]: Where is this parameter used?

File: playbooks/rhv/inventory.example [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+debug_level=2
```
Comment (sradco) [inline playbooks/rhv/inventory.example]: Checked and 2 is the default https://docs.okd.io/3.10/install/configuring_inventory_file.html#configuring-cluster-variables

File: playbooks/rhv/README.md [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+an OpenShift Container Platform environment on Red Hat Virtualization
+
+### Where do I start?
+Choose a host from which Ansible plays will be executed. This host must have
+the ability to access the web interface of the RHV cluster engine and the
+network on which the OpenShift nodes will be installed. We will refer to
+this host as the *bastion*.
+
+#### oVirt Ansible Roles
+The oVirt project maintains Ansible roles for maanaging an oVirt or RHV cluster.
```
Comment (machacekondra) [inline playbooks/rhv/README.md]: s/maanaging/managing

File: playbooks/rhv/README.md (hunk 14 above)
Comment (machacekondra) [reply playbooks/rhv/README.md]: It's not.

File: roles/openshift_rhv/tasks/build_vm_list.yml [hunk 21]
```diff
@@ -0,0 +1,41 @@
+---
+# Creates a dictionary for use with oVirt.vm-infra role
+# https://github.com/oVirt/ovirt-ansible-vm-infra
+- name: Create virtual machine list fact
```
Comment (machacekondra) [inline roles/openshift_rhv/tasks/build_vm_list.yml]: Why not add this support to `vm-infra` role?

File: roles/openshift_rhv/tasks/build_vm_list.yml [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+  set_fact:
+    vms: >-
+      {{ vms|default([]) }} + [
+      {% for iter in range(item.count) -%}
+      {% if iter > 0 -%},{% endif -%}
+      {
+      'name': '{{ item.name }}{{ iter }}.{{ public_hosted_zone }}',
```
Comment (sradco) [inline roles/openshift_rhv/tasks/build_vm_list.yml]: What var files do you use for this ? The public_hosted_zone in configured in the OpenShift variables file but not in the provisioning-vars.yaml.example

File: playbooks/rhv/README.md (hunk 2 above)
Comment (cwilkers) [reply playbooks/rhv/README.md]: @sradco That's not in line with the convention of the playbooks/{provider} structure in this repo.
Here, we're separating out provisioning from OpenShift install, thus the two files with vars in each that need editing.

File: playbooks/rhv/README.md (hunk 6 above)
Comment (cwilkers) [reply playbooks/rhv/README.md]: That seems to only apply to running from the engine host.

File: playbooks/rhv/inventory.example (hunk 9 above)
Comment (cwilkers) [reply playbooks/rhv/inventory.example]: +1, I'll take a pass at doing that today.

File: playbooks/rhv/inventory.example (hunk 11 above)
Comment (cwilkers) [reply playbooks/rhv/inventory.example]: That would be worth testing on (I've previously found bugs in load balancer configs WRT firewalling) but turning that on in the RHV example inventory contradicts our decision earlier in the thread to simplify the vars and inventory files.

File: playbooks/rhv/inventory.example [hunk 23]
```diff
@@ -0,0 +1,99 @@
+[all:vars]
+app_dns_prefix=apps
+public_hosted_zone=example.com
```
[the remark this answers is not in the document]
Comment (cwilkers) [reply playbooks/rhv/inventory.example]: +1 -- BUT, This is just a local convention for ease of use, and might be done away with if I'm paring down the vars.

File: playbooks/rhv/inventory.example (hunk 11 above)
Comment (sradco) [reply playbooks/rhv/inventory.example]: I don't see this flag in the https://github.com/openshift/openshift-ansible/blob/master/inventory/hosts.example ...
So either add it there or here no?

File: playbooks/rhv/inventory.example (hunk 12 above)
Comment (cwilkers) [reply playbooks/rhv/inventory.example]: Ah, I'll update to look like: https://github.com/openshift/openshift-ansible/blob/master/inventory/hosts.example#L39

File: playbooks/rhv/inventory.example (hunk 23 above)
[the remark this answers is not in the document]
Comment (sradco) [reply playbooks/rhv/inventory.example]: I think this is a mandatory variable to be set here by the user..

[the remark this answers is not in the document]
Comment (sradco) [reply playbooks/rhv/inventory.example]: Also app_dns_prefix which the user must set for his app.

[the remark this answers is not in the document]
Comment (sradco) [reply playbooks/rhv/inventory.example]: Not sure if app_dns_prefix is mandatory.
User can install OpenShift and add that later even in the ui right?

File: playbooks/rhv/inventory.example (hunk 15 above)
Comment (cwilkers) [reply playbooks/rhv/inventory.example]: `HTPasswdPasswordIdentityProvider` sets up an htpasswd file on all three masters, and you have to add an entry to that, my example being myuser with password changeme.

`AllowAllPasswordIdentityProvider` allows in any user with any userid/password combination.

Neither have anything specific to do with RHV, so I'd rather remove the section and leave it to the user to specify their own.

Comment (sradco) [reply playbooks/rhv/inventory.example]: Is there a default for openshift_master_identity_providers?  If not, I think we should give something and I'll add automation for it.

File: playbooks/rhv/provisioning-vars.yaml.example (hunk 16 above)
Comment (cwilkers) [reply playbooks/rhv/provisioning-vars.yaml.example]: They're mounted during prerequisites.yml, in the `container_runtime` role's `extra_storage_setup` task:
https://github.com/openshift/openshift-ansible/blob/master/roles/container_runtime/tasks/extra_storage_setup.yml

Vars set in master and node group vars: https://github.com/openshift/openshift-ansible/pull/9570/files#diff-8aa4dd60378793fbc8c2f195dec32c70R65

File: playbooks/rhv/openshift-cluster/register-nodes.yml (hunk 17 above)
Comment (cwilkers) [reply playbooks/rhv/openshift-cluster/register-nodes.yml]: I don't think that's necessary.

File: playbooks/rhv/inventory.example (hunk 18 above)
Comment (cwilkers) [reply playbooks/rhv/inventory.example]: That may(should) be the default. I'll check and remove if it is.

File: playbooks/rhv/openshift-cluster/register-nodes.yml (hunk 17 above)
Comment (sradco) [reply playbooks/rhv/openshift-cluster/register-nodes.yml]: ok, Will it fail if lb in undefined?

File: roles/openshift_rhv/tasks/build_vm_list.yml (hunk 21 above)
Comment (cwilkers) [reply roles/openshift_rhv/tasks/build_vm_list.yml]: +1 That sounds like a good idea.

In addition, some of the DNS automation code would be helpful here.

File: roles/openshift_rhv/tasks/build_vm_list.yml (hunk 22 above)
Comment (cwilkers) [reply roles/openshift_rhv/tasks/build_vm_list.yml]: Good catch. I'll need to better document that and ensure those vars are properly name-spaced to openshift_rhv.

File: playbooks/rhv/openshift-cluster/register-nodes.yml (hunk 17 above)
Comment (cwilkers) [reply playbooks/rhv/openshift-cluster/register-nodes.yml]: Actually, I'm going to remove this playbook; prerequisites.yml invalidates the need for it.

File: playbooks/rhv/README.md (hunk 14 above)
Comment (sradco) [reply playbooks/rhv/README.md]: @machacekondra Can't we add this step? @rgolangh Can you please share how you populate the ca.pem?

File: playbooks/rhv/openshift-cluster/register-nodes.yml (hunk 17 above)
Comment (sradco) [reply playbooks/rhv/openshift-cluster/register-nodes.yml]: +1

File: playbooks/rhv/inventory.example (hunk 11 above)
Comment (sradco) [reply playbooks/rhv/inventory.example]: But I don't see you added it to the openshift vars example file. Only for the oVirt playbooks..

File: playbooks/rhv/inventory.example [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+
+[OSEv3:children]
+nodes
+masters
+etcd
+lb
+
+[OSEv3:vars]
+# General variables
+ansible_ssh_user=root
```
Comment (sradco) [inline playbooks/rhv/inventory.example]: I saw you pushed some patches. This is still not updated.

File: playbooks/rhv/inventory.example [hunk 25]
```diff
@@ -0,0 +1,61 @@
+[all:vars]
```
Comment (sradco) [inline playbooks/rhv/inventory.example]: Why did you remove openshift_release?

File: playbooks/rhv/inventory.example (hunk 24 above)
Comment (cwilkers) [reply playbooks/rhv/inventory.example]: Thanks, the comment flow here makes for a horrible task list.

Comment (sradco) [reply playbooks/rhv/inventory.example]: Agree!

File: playbooks/rhv/inventory.example (hunk 25 above)
Comment (cwilkers) [reply playbooks/rhv/inventory.example]: I believe it defaults to current (pertinent for master branch), and it is not specific to RHV

File: playbooks/rhv/inventory.example [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+openshift_rhv_dns_zone=  # example.com
+
```
Comment (sradco) [inline playbooks/rhv/inventory.example]: I checked the default for openshift_master_identity_providers
https://docs.openshift.com/container-platform/3.9/install_config/configuring_authentication.html#identity-providers-ansible
and its Deny All.

It means that after OpenShift installation user will not be able to login to OpenShift?

Comment (cwilkers) [reply playbooks/rhv/inventory.example]: cluster:admin will still work from the master nodes, but no, the web interface requires some authentication setup in order to allow users in.

The goal of this PR is not to re-home the entire reference architecture in openshift-ansible, but to provide the relevant oVirt/RHV bits to help users install OCP on top of oVirt/RHV. Authentication documentation is better handled (and examples provided) elsewhere.

File: playbooks/rhv/README.md [hunk 27]
```diff
@@ -0,0 +1,79 @@
...
+#### Step 3 Provision Virtual Machines in RHV
+Once all the variables in the `provisioning_vars.yaml` file are set, use the
+[`ovirt-vm-infra.yml`](openshift-cluster/ovirt-vm-infra.yml) playbook to begin
+provisioning.
+
+```
+ansible-playbook -i inventory -e@provisioning_vars.yml ${PATH_TO_OPENSHIFT_ANSIBLE}/playbooks/rhv/openshift-cluster/ovirt-vm-infra.yml
+```
+
+Ensure that DNS is set up properly at this stage.
```
Comment (michaelgugino) [inline playbooks/rhv/README.md]: Missing step 4, looks like it should be a callout about dns.

File: playbooks/rhv/README.md [hunk 28]
```diff
@@ -0,0 +1,82 @@
...
+[`ovirt-vm-infra.yml`](openshift-cluster/ovirt-vm-infra.yml) playbook to begin
+provisioning.
+
+```
+ansible-playbook -i inventory -e@provisioning_vars.yml ${PATH_TO_OPENSHIFT_ANSIBLE}/playbooks/rhv/openshift-cluster/ovirt-vm-infra.yml
+```
+
+#### Step 4 Update DNS
+
+At this stage, ensure the newly provisioned nodes are reachable by thier
```
Comment (michaelgugino) [inline playbooks/rhv/README.md]: s/thier/their/

More important than the bastion host can contact nodes by their names is that nodes can resolve their own names, specifically the output of 'hostname -f' should be resolvable from each node.  Not sure how to best word this.

File: playbooks/rhv/README.md [hunk 29]
```diff
@@ -0,0 +1,82 @@
...
+hostnames, and the wildcard entry points at the infrastructure nodes.
```
Comment (michaelgugino) [inline playbooks/rhv/README.md]: I think in most production setups, the infra nodes (routers) are behind a load balancer.  Each infra node might not have a router, that is configurable, so need to ensure your documentation accounts for this (either here, or elsewhere, such as where you describe infra nodes).

**Comment by cwilkers** [comment]
Rebased and squashed...
